### PR TITLE
Draft: Add show-downstream, upstream-conflicts, graph-upstreams

### PR DESCRIPTION
**Pull request by yumaikas**
NOT ready to merge yet, still needs unit tests, but putting this up to keep track of this branch against main.

### CODE REVIEW

File: utils/query-state/Get-ConflictingBranches.psm1 [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+function Get-MergeConflicts([string]$a, [string]$b) {
+	$output = git merge-tree --name-only --write-tree $a $b 
+	$isClean = $LASTEXITCODE -eq 0 # git-merge-tree uses 0 for clean, 1 for not, and <0 for various other errors
+	return [ordered]@{
+		isClean = $isClean;
+		# Just because this list is empty doesn't mean there aren't conflicts
+		conflictList = ($output | Select-Object -Skip 1).Where({ $_ -eq "" }, [System.Management.Automation.WhereOperatorSelectionMode]::Until)
+	}
```
Comment (mdekrey) [inline utils/query-state/Get-ConflictingBranches.psm1:22]: To make the conflictList useful, you'll need `--no-messages` to be added. This should also use `Invoke-ProcessLogs`... I'm gonna make one in another PR that you'll be able to leverage here.

Comment (yumaikas) [reply utils/query-state/Get-ConflictingBranches.psm1:22]: This uses Where(Predicate, Until) to clear off the messages (though I don't know how git handles not having any messages)

But point taken

File: utils/query-state/Get-UpstreamBranchMap.psm1 [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+				path=$row[1] 
+			}
+		} `
+		| Where-Object { 
+			$branches -contains $_.path -or $branches.Length -eq 0 
+		};
+
+	$blobs = @{}
+	(git ls-tree -r origin/_upstream --format="%(objectname)" 
+		| git cat-file --batch="%(objectname) %(rest)")  -join "`n" -split "\n\n" `
```
Comment (mdekrey) [inline utils/query-state/Get-UpstreamBranchMap.psm1:17]: This is really cool, though it needs to leverage `Get-Configuration` to work in all the standard scenarios. This can be incorporated, for sure.

We can probably simplify this entire function a bit in order to only call the `git ls-tree` once... 🤔

File: utils/query-state/Select-UpstreamBranchGraph.psm1 [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		}
+		Write-Host ("=" * 24)
+		Write-Host
+	}
+}
+
+function Select-UpstreamBranchGraph([String]$branchName, [switch]$checkList=$false) {
+	$config = Get-Configuration;
+	if ($branchName -eq $config.defaultServiceLine) {
+		Write-Host "Cannot graph $($config.defaultServiceLine), it is the default service line and should not have upstreams";
```
Comment (mdekrey) [inline utils/query-state/Select-UpstreamBranchGraph.psm1:48]: FYI, service lines can have upstreams, just don't in anything simpler than the "Multiple maintained versions" flow. That is, `line/2.1` would have an upstream of `line/2.0` which would have an upstream of `line/1.4`, etc.

The default service line should just be the user's default; that's why it's not stored in a global config. (I could see a well-informed team using a "service line" for a big future release if it has thorough internal QA before merging into it.)

**Comment by mdekrey** [comment]
Closing this PR because a) it is out of date and b) related features have been addressed either by the [ScaledGitApp](https://github.com/PrincipleStudios/ScaledGitApp) or by newer commands within this repo.
